PayloadPkg/FirmwareUpdate/InternalFirmwareUpdate.c

```diff
@@ -0,0 +1,352 @@
+/** @file
+Internal functions to update firmware in boot media.
+
+Copyright (c) 2017, Intel Corporation. All rights reserved.<BR>
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <Base.h>
+#include <Uefi/UefiBaseType.h>
+#include <Library/PcdLib.h>
+#include <Library/BaseLib.h>
+#include <Library/DebugLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/BootloaderCommonLib.h>
+#include <Library/PayloadMemoryAllocationLib.h>
+#include <Library/FirmwareUpdateLib.h>
+#include <Library/ContainerLib.h>
+#include "InternalFirmwareUpdate.h"
+
+/**
+  Update a region block.
+
+  This is the acture function to update boot meia. It will erase boot device,
+  write new data to boot device, and verify the written data.
+
+  @param[in] Address          The boot media address to be update.
+  @param[in] Buffer           The source buffer to write to the boot media.
+  @param[in] Length           The length of data to write to boot media.
+
+  @retval  EFI_SUCCESS        Update successfully.
+  @retval  others             Error happening when updating.
+**/
+EFI_STATUS
+EFIAPI
+UpdateRegionLessthan4k (
+  IN  UINT64    Address,
+  IN  VOID      *Buffer,
+  IN  UINT32    Length
+  )
+{
+  EFI_STATUS    Status;
+  UINT8         ReadBuffer[0x1000];

```

Comment:
This array to too big to be on stack.  I think it might be better to allocate instead.

Reply:
Changed to allocate memory